### PR TITLE
Completed / Radu's Bowling-Challenge

### PR DESCRIPTION
**Pull request by necsanuradu**
The Game class has 25 methods explicitly named to declare their behavior/funtionality.
* 25 unit tests, testing for edge cases and behavior/functionality in one of the each: (Perfect / Gutter / Normal) game.
* The test coverage is 100%.

### CODE REVIEW

File: Gemfile [hunk 1]
```diff
@@ -0,0 +1,7 @@
+# frozen_string_literal: true
+
+source "https://rubygems.org"
+
+git_source(:github) { |repo_name| "https://github.com/#{repo_name}" }
+
+# gem "rails"
```
Comment (eballantine) [inline Gemfile:7]: If you're not using any gems, you don't need to run bundle. Gemfile and Gemfile.lock unnecessary bloat.

File: coverage/.last_run.json [hunk 2]
```diff
@@ -0,0 +1,5 @@
+{
+  "result": {
+    "line": 100.0
+  }
+}
```
Comment (eballantine) [inline coverage/.last_run.json:5]: git ignore this file

File: coverage/.resultset.json [hunk 3]
```diff
@@ -0,0 +1,291 @@
...
+          null,
+          null,
+          null,
+          null
+        ]
+      }
+    },
+    "timestamp": 1630327513
+  }
+}
```
Comment (eballantine) [inline coverage/.resultset.json:291]: similarly, can git ignore this file.

File: lib/game.rb [hunk 4]
```diff
@@ -0,0 +1,154 @@
+
+class Game
+    attr_reader :player, :frame, :roll, :score, :count_next, :pins_rest , :list_boundaries, :message
+
+    def initialize(player, pins_rest = 10, frame = 1, roll = 0, score = 0, count_next = 0)
+      @player, @frame, @roll, @this_roll_count, @this_roll_bonus = capitalize_name(player), frame, roll, 0, 0
```
Comment (eballantine) [inline lib/game.rb:6]: assigning so many variables on a single line like this is not easy to read, particularly when some of the variables have similar names/values, would be more readable over multiple lines.

File: lib/game.rb [hunk 5]
```diff
@@ -0,0 +1,154 @@
...
+      @score, @count_next, @pins_rest, @this_roll_messg =  score, count_next, pins_rest, nil
```
Comment (eballantine) [inline lib/game.rb:7]: by comparison, this line is easier to read.

File: lib/game.rb [hunk 6]
```diff
@@ -0,0 +1,154 @@
...
+      check_initialize_arguments
+    end
+
+    def check_initialize_arguments
+      @list_boundaries.each do |arg, max| 
+        raise 'player has non-number ' << arg.to_s unless eval('@' << arg.to_s).is_a?(Integer) 
+        raise 'player has negative ' << arg.to_s if eval('@' << arg.to_s) < 0 
+        raise 'player has ' << arg.to_s << ' greater than ' << max.to_s if eval('@' << arg.to_s) > max 
+      end
+    end
```
Comment (eballantine) [inline lib/game.rb:18]: This method is great, excellent safety net to protect the integrity of the game.

File: lib/game.rb [hunk 7]
```diff
@@ -0,0 +1,154 @@
...
+
+    def register_pins(pins_number = 0)
```
Comment (eballantine) [inline lib/game.rb:20]: I feel like it doesn't make sense to set a default value for pins. You would not want to register a bowl of 0 unless this is specified.

File: lib/game.rb [hunk 8]
```diff
@@ -0,0 +1,154 @@
...
+      raise 'pins registration failed, game has ended' if @frame == 10 && @roll == 2 && @count_next == 0
+      raise 'pins registration failed, game has ended' if @frame == 10 && @roll == 3
+    end
+
+    def check_pins_number
+      raise 'pins registration failed, not a number' unless @pins_number.is_a?(Integer)
+      raise 'pins registration failed, number smaller than 0' if @pins_number < 0
+      raise 'pins registration failed, number greater than 10' if @pins_number > 10
+      raise 'pins registration failed, number greater than rest of pins' if @pins_number > @pins_rest
+    end
```
Comment (eballantine) [inline lib/game.rb:146]: Another good method protecting the integrity of the game.

File: lib/game.rb [hunk 9]
```diff
@@ -0,0 +1,154 @@
...
+
+    def plus_roll_bonus
+      return unless bonus_roll?
+      @score += @pins_number
+      @count_next -= 1
+      @this_roll_bonus = @pins_number
+    end
+
+    def plus_bonus_on_top_of_bonus
```
Comment (eballantine) [inline lib/game.rb:130]: unclear method name

File: spec/unit/game_spec.rb [hunk 10]
```diff
@@ -0,0 +1,122 @@
...
+              }
+
+  let(:subject){ described_class.new(player.name) }
+  let(:subject_4_left){ described_class.new(player_4_left.name, player_4_left.pins_rest) }
+  let(:subject_game_has_ended){ described_class.new(player_end.name, player_end.pins_rest, player_end.frame, player_end.roll, player_end.score, player_end.count_next) }
+  let(:subject_game_has_extra_roll){ described_class.new(player_end.name, player_end.pins_rest, player_end.frame, player_end.roll, player_end.score, player_extra.count_next) }
+
+  let(:list){ ['player', 'frame', 'roll', 'score', 'count_next'] }
+
+  it 'expects the instantiated class should hold these state instance variables' do
```
Comment (eballantine) [inline spec/unit/game_spec.rb:36]: Which state instance variables? Reading just this test description, it is not clear what to expect.

File: spec/unit/game_spec.rb [hunk 11]
```diff
@@ -0,0 +1,122 @@
...
+    expect{ subject.register_pins(11) }.to raise_error('pins registration failed, number greater than 10')
+  end
+
+  it 'expects register_pins to raise error if argument(dropped pins number) is greater than pins_rest(left standing) ' do
+    expect{ subject_4_left.register_pins(5) }.to raise_error('pins registration failed, number greater than rest of pins')
+  end
+  
+  it 'expects register_pins to raise error if game has ended ' do
+    expect{ subject_game_has_ended.register_pins(3) }.to raise_error('pins registration failed, game has ended')
+  end
```
Comment (eballantine) [inline spec/unit/game_spec.rb:64]: The 5 tests above, well written, clear what you're testing.

File: spec/unit/game_spec.rb [hunk 12]
```diff
@@ -0,0 +1,122 @@
...
+
+  # it 'expects register_pins to register pins having an extra round' do
+  #   expect(subject_game_has_extra_roll.register_pins(3)).to eq('extra roll')
+  # end
```
Comment (eballantine) [inline spec/unit/game_spec.rb:68]: remove unused code before PR

File: spec/unit/game_spec.rb [hunk 13]
```diff
@@ -0,0 +1,122 @@
...
+      end
+    end
+  end
+
+  context "check all values are not greater than maximum allowed" do
+    player_name = "Jake"
+    described_class.new(player_name).list_boundaries.each do |arg, max|
+      it 'expects subject to raise error when initialized argument greater than #{max}' do
+        allow(player).to receive(arg).and_return(max + 1)
+        expect{ described_class.new(player.name, player.pins_rest, player.frame, player.roll, player.score, player.count_next) }.to raise_error('player has ' << arg.to_s << ' greater than ' << max.to_s)
```
Comment (eballantine) [inline spec/unit/game_spec.rb:93]: this (and lines 74, 83) too long!

**Review by eballantine — COMMENTED** [review]
Very thorough, particularly guarding for invalid inputs. 

Good test suite, checking for normal gameplay and for expected handling of invalid inputs.

As a reviewer, it is hard to read some of your code due to the naming of variables and methods being similar and/or not specific.

**Review by necsanuradu — COMMENTED** [review]
Thank you Lizzy

File: Gemfile (hunk 1 above)
Comment (necsanuradu) [reply Gemfile:7]: True, thank you

File: coverage/.last_run.json (hunk 2 above)
Comment (necsanuradu) [reply coverage/.last_run.json:5]: True, thank you

File: lib/game.rb (hunk 4 above)
Comment (necsanuradu) [reply lib/game.rb:6]: Absolutely

File: lib/game.rb (hunk 6 above)
Comment (necsanuradu) [reply lib/game.rb:18]: Thank you

File: lib/game.rb (hunk 7 above)
Comment (necsanuradu) [reply lib/game.rb:20]: I thought about the situation of fault or miss when we still have to move with the frame.

File: lib/game.rb (hunk 8 above)
Comment (necsanuradu) [reply lib/game.rb:146]: Thank you

File: lib/game.rb (hunk 9 above)
Comment (necsanuradu) [reply lib/game.rb:130]: Right again, thank you

File: spec/unit/game_spec.rb (hunk 11 above)
Comment (necsanuradu) [reply spec/unit/game_spec.rb:64]: Thank you

File: spec/unit/game_spec.rb (hunk 13 above)
Comment (necsanuradu) [reply spec/unit/game_spec.rb:93]: Absolutely, thank you
